Tweak the EDD doc (#184)
bitwarden/contributing-docs: docs/contributing/database-migrations/edd.mdx

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";

# Evolutionary database design

At Bitwarden we follow [Evolutionary Database Design (EDD)][edd-wiki]. EDD describes a process where
the database schema is continuously updated while still ensuring compatibility with older releases
by defining a database transition phases.

> Bitwarden also need to support:

Reviewer: ⛏️ "needs".

## Design

Database changes can be categorized into two categories: destructive and non-destructive changes

> \[[1](./edd#further-reading)\]. A destructive change prevents existing functionality from working as

Reviewer: 🎨 Can we link this to wording vs. a reference tag? Could link "two categories".

Author: It feels more like a foot note since the categories are denoted below and a link could be presumed to go to the categories within the document.

expected without an accompanying code change. A non-destructive change is the opposite: a database
change that does not require a code change to allow the non-application to continue working as
expected.

### Non-destructive changes

Many database changes can be designed in a backwards compatible manner by using a mix of nullable
fields and default values in the database tables, views, and stored procedures. This ensures that

> the stored procedures can be called without the new columns which allows it to run with both the old

Reviewer: ⛏️ Replace "which allows it" with "and allow them".

and new code.

### Destructive changes

Any change that cannot be done in a non-destructive manner is a destructive change. This can be as
simple as adding a non nullable column where the value needs to be computed from existing fields, or
renaming an existing column. To handle destructive changes it's necessary to break them up into

> three phases: _Start_, _Transition_ and _End_ as shown in the diagram below.

Reviewer: ⛏️ Oxford comma after _Transition_.

<figure>

![Refactoring Stages](./transitions.png)

<figcaption>Refactoring Phases</figcaption>

</figure>

It's worth noting that the _Refactoring Phases_ are usually rolling, and the _End phase_ of one
refactor is the _Transition phase_ of another. The table below details which application releases
needs to be supported during which database phase.

| Database Phase | Release X | Release X+1 | Release X+2 |
| -------------- | --------- | ----------- | ----------- |
| Start          | ✅        | ❌          | ❌          |
| Transition     | ✅        | ✅          | ❌          |
| End            | ❌        | ✅          | ✅          |

### Migrations

The three different migrations described in the diagram above are, _Initial migration_, _Transition

> migration_ and _ Finalization migration_ and are described in more details in this section.

Reviewer: ⛏️ Can drop "and are ..." as it feels implied.

#### Initial migration

> The initial migration runs before the code deployment, and it's purpose is to add support for

Reviewer: ⛏️ "its".

> _Release X+1_ without breaking support of _Release X_. This deployment should have execute quickly

Reviewer 2: ```suggestion
_Release X+1_ without breaking support of _Release X_. The migration should execute quickly
```

and not contain any costly operations to ensure zero downtime.

#### Transition migration

> The transition migration are run sometime during the transition phase, and provides an optional data

Reviewer: ⛏️ "migrations".

Author: I view this from the perspective of a single "change". In that scope it's singular migrations. However a deployment may contain multiple changes, and therefore multiple migrations.

> migration should it be to slow or put to much load on the database, or otherwise making it

Reviewer: ⛏️ "too" in two cases.

> migration should it be to slow or put to much load on the database, or otherwise making it

Reviewer: ⛏️ "make".

> unsuitable for the _initial migration_.

Reviewer: ⛏️ Capitalize _Initial_.

- Compatible with _Release X_ **and** _Release X+1_ application.
- Only data population migrations may be run at this time, if they are needed
  - Must be run as a background task during the Transition phase.
  - Operation is batched or otherwise optimized to ensure the database stays responsive.
- Schema changes are NOT to be run during this phase.

#### Finalization migration

> The finalization migration removes the temporary measurements that was needed to retain backwards

Reviewer: ⛏️ "were".
